src/networking.c

```diff
@@ -3390,6 +3390,21 @@ int checkClientOutputBufferLimits(client *c) {
     if (server.client_obuf_limits[class].hard_limit_bytes &&
         used_mem >= server.client_obuf_limits[class].hard_limit_bytes)
         hard = 1;
+    /* Since master may copy almost whole replication backlog to replica's
+     * output buffer if it is able to perform a partial resynchronization,
+     * but we will close replica client and resynchronization is failed if
+     * output buffer limit is reached. This situation might continue until
+     * replication backlog start offset is out of the PSYNC offset, and
+     * causes full synchronization.
+     * To avoid above bad situation, if the replica output buffer size is not
+     * more than replication backlog size but already exceeds output buffer
+     * hard limit, we don't think hard limit is reached and don't close replica
+     * client. This method doesn't cost more memory since replication backlog
+     * and replicas share one global replication buffer. */
+    if (hard == 1 && class == CLIENT_TYPE_SLAVE &&
+        used_mem <= (unsigned long)server.repl_backlog_size)
+        hard = 0;
```

Comment:
i thought we'll do this: (i think it's also cleaner)
```c
size_t hard_limit_bytes = server.client_obuf_limits[class].hard_limit_bytes;
if (class == CLIENT_TYPE_SLAVE && hard_limit_bytes < server.repl_backlog_size)
    hard_limit_bytes = server.repl_backlog_size;
```
then just use hard_limit_bytes instead of `server.client_obuf_limits[class].hard_limit_bytes`

WDYT?